### PR TITLE
implemented feature for unicode range

### PR DESCRIPTION
**Pull request by karina-usmanova**

### CODE REVIEW

File: pire/extra/unicode_range.cpp [hunk 1]
```diff
@@ -0,0 +1,134 @@
...
+            }
+
+            if (ch != ']') {
+                Error("Pire::UnicodeRangeImpl::Lex(): \"\\r[...\" sequence should be closed by \"]\"");
+            }
+
+            return Term(TokenTypes::Letters, range);
+        }
+    private:
+        const wchar32 MAX_UNICODE = 0x10FFFF;
```
Comment (sergey-v-galtsev) [inline pire/extra/unicode_range.cpp]: IMHO
`static const ...`

Style
PIRE's constants are written mostly in CamelCase.

File: pire/extra/unicode_range.cpp [hunk 2]
```diff
@@ -0,0 +1,134 @@
...
+
+        bool IsHexDigit(wchar32 ch) {
+            return ch < 256 && std::isxdigit(ch) != 0;
+        }
+
+        ystring ReadHexDigit(std::function<bool(wchar32, size_t)> shouldStop) {
```
Comment (sergey-v-galtsev) [inline pire/extra/unicode_range.cpp]: IMHO
It looks like a copy-paste of `pire/extra/unicode_support.cpp`.
So it is better to extract this code to a separate file and include it from `unicode_support.cpp` and from `unicode_range.cpp`.

File: pire/extra/unicode_range.cpp [hunk 3]
```diff
@@ -0,0 +1,134 @@
...
+namespace Pire {
+
+namespace {
+    class UnicodeRangeImpl : public Feature {
+    public:
+        bool Accepts(wchar32 c) const {
+            return c == (Control | 'r');
+        }
+
+        Term Lex() {
```
Comment (sergey-v-galtsev) [inline pire/extra/unicode_range.cpp]: Please, add some tests.
E.g.
- various erroneous regexps,
- regexps with one simple range,
- regexps with a negated range,
- regexps with several ranges inside `\r[...]`.

**Comment by sergey-v-galtsev** [comment]
Looks good to me.

File: tests/read_unicode_ut.cpp [hunk 4]
```diff
@@ -0,0 +1,272 @@
...
+		REGEXP("[^\\x{61}A]") {
+			ACCEPTS("b");
+			ACCEPTS("B");
+			ACCEPTS("atestB");
+			DENIES("a");
+			DENIES("A");
+			DENIES("aaAA");
+		}
+
+		REGEXP("[0-9][\\x{61}-\\x{62}A-B]") {
```
Comment (sergey-v-galtsev) [inline tests/read_unicode_ut.cpp]: IMHO
It would be great to test mixing of code and letter in the same range also.
Eg.: `[\\x{61}-bA-\\x{42}]`

**Comment by sergey-v-galtsev** [comment]
Looks good to me.

**Comment by kv75** [comment]
Looks good to me

**Pull request merged**